core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java

```diff
@@ -510,6 +510,118 @@ private void basePushFilterPastAggWithGroupingSets(boolean unchanged)
     checkPlanning(tester, preProgram, new HepPlanner(program), sql);
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-1338">[CALCITE-1338]
+   * Should not transpose since the right project of left join has literal</a>. */
+  @Test public void testJoinProjectTranspose2() {
+    final HepProgram program =
+        HepProgram.builder()
+            .addRuleInstance(JoinProjectTransposeRule.RIGHT_PROJECT_INCLUDE_OUTER)
+            .build();
+    final String sql = "select *\n"
+        + "from dept a \n"
+        + "left join (select name, 1 from dept) as b\n"
+        + "on a.name = b.name";
+    checkPlanUnchanged(new HepPlanner(program), sql);
+  }
+
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-1338">[CALCITE-1338]
+   * Should not transpose since the left project of right join has literal</a>. */
+  @Test public void testJoinProjectTranspose3() {
+    final HepProgram program =
+        HepProgram.builder()
+            .addRuleInstance(JoinProjectTransposeRule.LEFT_PROJECT_INCLUDE_OUTER)
+            .build();
+    final String sql = "select *\n"
+        + "from (select name, 1 from dept) as a\n"
+        + "right join dept b\n"
+        + "on a.name = b.name";
+    checkPlanUnchanged(new HepPlanner(program), sql);
+  }
+
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-1338">[CALCITE-1338]
+   * Should not transpose since the right project of left join has not-strong
+   * expression {@code y is not null}</a>. */
+  @Test public void testJoinProjectTranspose4() {
+    final HepProgram program =
+        HepProgram.builder()
+            .addRuleInstance(JoinProjectTransposeRule.RIGHT_PROJECT_INCLUDE_OUTER)
+            .build();
+    final String sql = "select *\n"
+        + "from dept a\n"
+        + "left join (select x name, y is not null from\n"
+        + "(values (2, cast(null as integer)), (2, 1)) as t(x, y)) b\n"
+        + "on a.name = b.name";
+    checkPlanUnchanged(new HepPlanner(program), sql);
```

Comment:
Hi @chunweilei , I'm learning this PR ~
In the right project of this left-join, the field of `name` is of Integer type. Do you really mean this?